### PR TITLE
Sgdr fix

### PR DESCRIPTION
**Pull request by PatrykChrabaszcz**
Proposed fix for SGDR implementation.
Developed together with the author @loshchil

### CODE REVIEW

**Comment by loshchil** [comment]
Dear Vijay Vasudevan @vrv ,

Please consider to review this request given the following 
i) A new function linear_cosine_decay recently added to TensorFlow implements some features (minimum learning rate and restarts) that for an unknown reason are missing in cosine_decay function implemented by the same TensorFlow contributor. Note that these features **were proposed in SGDR and also implemented here https://github.com/tensorflow/tensorflow/pull/11749 in TensorFlow *before* linear_cosine_decay was implemented**. Thus, the contributor of linear_cosine_decay function didn't take into account both the existing code and academic sources.
ii) You reviewed and accepted our implementation of SGDR  https://github.com/tensorflow/tensorflow/pull/11749 in TensorFlow  which preceded the one where linear_cosine_decay was introduced.
iii) You are a co-author of the paper where linear cosine decay was introduced.
iv) The current implementation of linear_cosine_decay misleads the users of TensorFlow to think that the use of minimum learning rate and restarts together with cosine decay was introduced by linear_cosine_decay function and the corresponding paper.

Please consider our fix designed to avoid misleading TensorFlow users from industry and academia.

Thank you for your consideration.
Ilya

**Comment by vrv** [comment]
This looks good to me.  If you want to have the cosine_decay_restarts in the public API , you'll need to also add the call out here: https://github.com/tensorflow/tensorflow/blob/e9d2b60ed3d94eef7a3cf139cc27ec629f510681/tensorflow/python/training/training.py#L41, and then I'll have to ask for someone from the API review team to approve the change as well.

Do you feel strongly about the new parameter name 'first_decay_steps' ?  In the contrib version, it is named differently. 

Lastly, I would also change the contrib version of the code to call this public version so there's not two copies of the implementation in the codebase that are largely the same.

Otherwise, looks good, and thanks for contributing!

**Comment by PatrykChrabaszcz** [comment]
We used the name **initial_period_steps** . We changed the name to **first_decay_steps** because  **linear_cosine_decay** and **cosine_decay** functions have a parameter **decay_steps** . We thought that this way we will keep the naming more consistent. I'm not sure which name is better, the current one or the previous one.

**Comment by vrv** [comment]
Okay, sounds good.  Let's proceed with adding the callouts in the public API then?

**Review by vrv — APPROVED** [review]

**Comment by vrv** [comment]
(I believe API review was skipped last week due to NIPS conference, I'm optimistic it'll happen this week).

File: tensorflow/contrib/training/python/training/sgdr_learning_rate_decay.py [hunk 1]
```diff
@@ -125,63 +120,6 @@ def sgdr_decay(learning_rate, global_step, initial_period_steps,
...
-        lambda: compute_step(x_val, geometric=False),
-        lambda: compute_step(x_val, geometric=True))
-
-    # If m_mul < 1, then the initial learning rate of every new restart will be
-    # smaller, i.e., by a factor of m_mul ** i_restart at i_restart-th restart
-    m_fac = learning_rate * (m_mul ** i_restart)
-
-  return math_ops.multiply(c_half * m_fac,
-                           (math_ops.cos(x_val * c_pi) + c_one), name=name)
+  return learning_rate_decay.cosine_decay_restarts(learning_rate, global_step,
```
Comment (taion) [inline tensorflow/contrib/training/python/training/sgdr_learning_rate_decay.py]: fwiw, since this was never added to https://github.com/tensorflow/tensorflow/blob/v1.4.1/tensorflow/contrib/training/__init__.py, it wasn't ever part of the publicly visible TF API (was not available via `tf.contrib.training.sgdr_decay`), so maybe it would be fine to remove this entirely?

Comment (PatrykChrabaszcz) [reply tensorflow/contrib/training/python/training/sgdr_learning_rate_decay.py]: I pushed a new commit with those 2 files removed.

**Review by asimshankar — COMMENTED** [review]
(For TF api owners):

Two questions that we defer to @vrv, otherwise okay:

(a) Would it make sense to remove linear_cosine_decay and instead have this one function that is a superset? (Note that since `linear_cosine_decay` isn't part of the last release (1.4) - we have a short window to remove it before the next release).

(b) Would it make sense to fold this behavior into `cosine_decay` (where by default there will be no restarts) instead of adding a new symbol?

Either way, defer to @vrv's judgement on this. Otherwise, good for API change.

Thanks!

**Comment by vrv** [comment]
Thanks Asim!

I like the idea of fewer public symbols, however, I am not sure whether cosine_decay should default to having restarts enabled or disabled.  The warm restarts idea was a big contribution of the original paper, but I kind of preferred the simplicity of removing hyperparameters that cosine_decay currently has.  I figured having two public APIs, one simple, one complex might be preferred.

Similarly, with linear_cosine_decay, etc; if we add num_periods and beta and other fields to cosine_decay, the documentation and code gets more complicated, and I thought it might make sense to keep things simple.

What do you all think?

**Comment by asimshankar** [comment]
@vrv : Sounds good. Rule of thumb: If it gets hard to explain the interaction between the arguments, then yeah, separate functions seem like a good idea. Thanks.

**Comment by vrv** [comment]
I think it might be fine to just keep them as separate functions, though they could probably share implementations more in the future.

I think you just need to resolve the conflicts in the file and we can get this in.

**Comment by drpngx** [comment]
@PatrykChrabaszcz fee free to pull rebase and push again.

**Comment by PatrykChrabaszcz** [comment]
@drpngx 
It should be done if I did not mess something up (first time I do this).

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
@PatrykChrabaszcz could you look into the build failures?

**Comment by PatrykChrabaszcz** [comment]
@drpngx I didn't remove one line that uses old "sgdr_learning_rate_decay.py" file from the BUILD file, fixed now

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Right, you need to update the goldens on the API. This adds a new operator in `tf.train`.

**Comment by PatrykChrabaszcz** [comment]
@drpngx I'm afraid that I do not understand what "update the goldens on the API" means :) . I see that some checks for Python2 fail because of API changes. Is there something I should do now to fix it?

**Comment by vrv** [comment]
@PatrykChrabaszcz If you look at the logs here, you will see instructions about how to run the api compatibility golden API generation files, and then you can include the changed files in a subsequent push.

https://source.cloud.google.com/results/invocations/08c9e185-a549-411a-9b11-0394502a8d7b/targets/%2F%2Ftensorflow%2Ftools%2Fapi%2Ftests:api_compatibility_test/log

Let us know if that's not helpful enough!

@drpngx perhaps we should have these instructions on CONTRIBUTIONS.md somewhere, rather than buried in a test file log.

**Comment by drpngx** [comment]
/CC @MarkDaoust @gunan +1 @vrv can't hurt to have it in the contributions doc.

**Comment by PatrykChrabaszcz** [comment]
@vrv 
I had some errors while compiling with the first command.
I had to add some additional options:
bazel build --copt=-msse4.1 --copt=-msse4.2 tensorflow/tools/api/tests:api_compatibility_test

I hope everything is ok now

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Timeout on `//tensorflow/python/keras:data_utils_test`, trying again.
